Gora 409

djkevincr opened the pull request.

lewismc commented on gora-compiler/src/main/java/org/apache/gora/compiler/GoraCompiler.java:
```diff
@@ -273,4 +274,8 @@ private static Schema getRecordSchemaWithDirtySupport(Schema originalSchema, Map
     return newSchema;
   }
 
+  public static long fingerprint64(Schema schema) {
```
Can you please document this method

lewismc commented on gora-compiler/src/main/velocity/org/apache/gora/compiler/templates/record.vm:
```diff
@@ -353,6 +353,33 @@ public class ${this.mangle($schema.getName())}#if ($schema.isError()) extends or
+          DATUM_READER$ = new org.apache.avro.specific.SpecificDatumReader(SCHEMA$);
+
+  @Override
+  public void writeExternal(java.io.ObjectOutput out)
```
Possible to document this method template?

lewismc commented on gora-compiler/src/main/velocity/org/apache/gora/compiler/templates/record.vm:
```diff
@@ -353,6 +353,33 @@ public class ${this.mangle($schema.getName())}#if ($schema.isError()) extends or
+  }
+
+  @Override
+  public void readExternal(java.io.ObjectInput in)
```
Possible to document this method template?

lewismc commented on gora-core/src/main/java/org/apache/gora/persistency/impl/DirtyCollectionWrapper.java:
```diff
@@ -1,3 +1,21 @@
+/**
```
Thank you for adding license header here.

lewismc commented on gora-core/src/main/java/org/apache/gora/persistency/impl/DirtyIteratorWrapper.java:
```diff
@@ -1,3 +1,21 @@
+/**
```
Thank you for adding license header here.

lewismc commented on gora-core/src/main/java/org/apache/gora/persistency/impl/DirtyListIterator.java:
```diff
@@ -1,3 +1,21 @@
+/**
```
Thank you for adding license header here.

lewismc commented on gora-core/src/main/java/org/apache/gora/persistency/impl/DirtyListWrapper.java:
```diff
@@ -1,3 +1,21 @@
+/**
```
Thank you for adding license header here.

lewismc commented on gora-core/src/main/java/org/apache/gora/persistency/impl/DirtyMapWrapper.java:
```diff
@@ -1,3 +1,21 @@
+/**
```
Thank you for adding license header here.

lewismc commented on gora-core/src/main/java/org/apache/gora/persistency/impl/DirtySetWrapper.java:
```diff
@@ -1,3 +1,21 @@
+/**
```
Thank you for adding license header here.

lewismc commented on gora-core/src/main/java/org/apache/gora/persistency/impl/PersistentBase.java:
```diff
@@ -181,10 +181,14 @@ public void setDirty(String field) {
   }
 
-  private ByteBuffer getDirtyBytes() {
+  public ByteBuffer getDirtyBytes() {
```
Can you document this please.

lewismc commented on gora-core/src/main/java/org/apache/gora/persistency/impl/PersistentBase.java:
```diff
@@ -181,10 +181,14 @@ public void setDirty(String field) {
     return __g__dirty;
   }
 
+  public void setDirtyBytes(ByteBuffer __g__dirty) {
```
Can you document this please.

lewismc commented on gora-core/src/main/java/org/apache/gora/store/DataStoreFactory.java:
```diff
@@ -42,6 +42,8 @@
 
   public static final String GORA_DEFAULT_DATASTORE_KEY = "gora.datastore.default";
 
+  public static final String GORA_DEFAULT_CACHE_DATASTORE_KEY = "gora.cache.datastore.default";
```
Can you document this please.

lewismc commented on gora-core/src/main/java/org/apache/gora/store/DataStoreFactory.java:
```diff
@@ -274,6 +276,39 @@ D getDataStore( Class<D> dataStoreClass, Class<K> keyClass,
+   *
+   * @param keyClass The key class.
+   * @param persistent The value class.
+   * @param conf {@link Configuration} to be used be the store.
```
Please use the package name for Configuration or else Javadoc may not work.

lewismc commented on gora-core/src/main/java/org/apache/gora/store/DataStoreFactory.java:
```diff
@@ -274,6 +276,39 @@ D getDataStore( Class<D> dataStoreClass, Class<K> keyClass,
+   * @param conf {@link Configuration} to be used be the store.
+   * @param isCacheEnabled caching enable
+   * @return A new store instance.
+   * @throws GoraException
```
If what happens? Why would it throw a GoraException?

lewismc commented on gora-core/src/main/java/org/apache/gora/store/DataStoreFactory.java:
```diff
@@ -381,6 +416,10 @@ private static String getDefaultDataStore(Properties properties) {
     return getProperty(properties, GORA_DEFAULT_DATASTORE_KEY);
   }
 
+  private static String getDefaultCacheDataStore(Properties properties) {
```
Add Javadoc here please.

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/query/JCacheQuery.java:
```diff
@@ -0,0 +1,34 @@
+import org.apache.gora.query.impl.QueryBase;
+import org.apache.gora.store.DataStore;
+
+public class JCacheQuery<K, T extends PersistentBase> extends QueryBase<K, T> {
```
Please document this class

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/query/JCacheQuery.java:
```diff
@@ -0,0 +1,34 @@
+
+public class JCacheQuery<K, T extends PersistentBase> extends QueryBase<K, T> {
+
+  public JCacheQuery() {
```
Please document this default constructor

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/query/JCacheQuery.java:
```diff
@@ -0,0 +1,34 @@
+    super(null);
+  }
+
+  public JCacheQuery(DataStore<K, T> dataStore) {
```
Please document this constructor

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/query/JCacheResult.java:
```diff
@@ -0,0 +1,69 @@
+import org.apache.gora.query.impl.ResultBase;
+import org.apache.gora.store.DataStore;
+
+public class JCacheResult<K, T extends PersistentBase> extends ResultBase<K, T> {
```
Please document this class

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/query/JCacheResult.java:
```diff
@@ -0,0 +1,69 @@
+  private NavigableSet<K> cacheKeySet;
+  private Iterator<K> iterator;
+
+  public JCacheResult(DataStore<K, T> dataStore, Query<K, T> query) {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/query/JCacheResult.java:
```diff
@@ -0,0 +1,69 @@
+    super(dataStore, query);
+  }
+
+  public JCacheResult(DataStore<K, T> dataStore, Query<K, T> query, NavigableSet<K> cacheKeySet) {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/query/JCacheResult.java:
```diff
@@ -0,0 +1,69 @@
+    this.iterator = cacheKeySet.iterator();
+  }
+
+  public JCacheStore<K, T> getDataStore() {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/query/JCacheResult.java:
```diff
@@ -0,0 +1,69 @@
+  }
+
+  @Override
+  public float getProgress() throws IOException {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/query/JCacheResult.java:
```diff
@@ -0,0 +1,69 @@
+  }
+
+  @Override
+  public void close() throws IOException {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/query/JCacheResult.java:
```diff
@@ -0,0 +1,69 @@
+  }
+
+  @Override
+  protected boolean nextInner() throws IOException {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheEntryListener.java:
```diff
@@ -0,0 +1,77 @@
+import javax.cache.event.CacheEntryListenerException;
+import java.util.concurrent.ConcurrentSkipListSet;
+
+public class JCacheCacheEntryListener<K, T extends PersistentBase>
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheEntryListener.java:
```diff
@@ -0,0 +1,77 @@
+  private static final Logger LOG = LoggerFactory.getLogger(JCacheCacheEntryListener.class);
+  private ConcurrentSkipListSet<K> cacheEntryList;
+
+  public JCacheCacheEntryListener(ConcurrentSkipListSet cacheEntryList) {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheEntryListener.java:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  @Override
+  public void onCreated(Iterable<CacheEntryEvent<? extends K, ? extends T>> cacheEntryEvents)
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheEntryListener.java:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  @Override
+  public void onRemoved(Iterable<CacheEntryEvent<? extends K, ? extends T>> cacheEntryEvents)
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheEntryListener.java:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  @Override
+  public void onUpdated(Iterable<CacheEntryEvent<? extends K, ? extends T>> cacheEntryEvents)
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheEntryListener.java:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  @Override
+  public void onExpired(Iterable<CacheEntryEvent<? extends K, ? extends T>> cacheEntryEvents)
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheEntryListenerFactory.java:
```diff
@@ -0,0 +1,56 @@
+
+import javax.cache.configuration.Factory;
+
+public class JCacheCacheEntryListenerFactory <K,T extends PersistentBase>
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheEntryListenerFactory.java:
```diff
@@ -0,0 +1,56 @@
+  public static final long serialVersionUID = 201305101634L;
+  private transient JCacheCacheEntryListener<K, T> instance;
+
+  public JCacheCacheEntryListenerFactory(JCacheCacheEntryListener<K, T> instance) {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheEntryListenerFactory.java:
```diff
@@ -0,0 +1,56 @@
+    this.instance = instance;
+  }
+
+  public JCacheCacheEntryListener<K, T> create() {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheEntryListenerFactory.java:
```diff
@@ -0,0 +1,56 @@
+    return this.instance;
+  }
+
+  public boolean equals(Object other) {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheEntryListenerFactory.java:
```diff
@@ -0,0 +1,56 @@
+    }
+  }
+
+  public int hashCode() {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheFactoryBuilder.java:
```diff
@@ -0,0 +1,41 @@
+import org.apache.gora.store.DataStore;
+import javax.cache.configuration.Factory;
+
+public class JCacheCacheFactoryBuilder {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheFactoryBuilder.java:
```diff
@@ -0,0 +1,41 @@
+
+public class JCacheCacheFactoryBuilder {
+
+  public static <K, T extends PersistentBase> Factory<JCacheCacheLoader<K,T>>
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheFactoryBuilder.java:
```diff
@@ -0,0 +1,41 @@
+    return new JCacheCacheLoaderFactory<>(new JCacheCacheLoader<>(dataStore));
+  }
+
+  public static <K, T extends PersistentBase> Factory<JCacheCacheWriter<K,T>>
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheFactoryBuilder.java:
```diff
@@ -0,0 +1,41 @@
+    return new JCacheCacheWriterFactory<>(new JCacheCacheWriter<>(dataStore));
+  }
+
+  public static <K,T extends PersistentBase> Factory<JCacheCacheEntryListener<K, T>>
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheLoader.java:
```diff
@@ -0,0 +1,65 @@
+import java.util.Map;
+
+
+public class JCacheCacheLoader<K, T extends PersistentBase> implements CacheLoader<K, T> {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheLoader.java:
```diff
@@ -0,0 +1,65 @@
+  private static final Logger LOG = LoggerFactory.getLogger(JCacheCacheLoader.class);
+  private DataStore<K, T> dataStore;
+
+  public JCacheCacheLoader(DataStore<K, T> dataStore) {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheLoader.java:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  @Override
+  public T load(K key) throws CacheLoaderException {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheLoader.java:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  @Override
+  public Map<K, T> loadAll(Iterable<? extends K> keys) throws CacheLoaderException {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheLoaderFactory.java:
```diff
@@ -0,0 +1,56 @@
+
+import javax.cache.configuration.Factory;
+
+public class JCacheCacheLoaderFactory<K, T extends PersistentBase>
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheLoaderFactory.java:
```diff
@@ -0,0 +1,56 @@
+  public static final long serialVersionUID = 201305101626L;
+  private transient JCacheCacheLoader<K, T> instance;
+
+  public JCacheCacheLoaderFactory(JCacheCacheLoader<K, T> instance) {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheLoaderFactory.java:
```diff
@@ -0,0 +1,56 @@
+    this.instance = instance;
+  }
+
+  public JCacheCacheLoader<K, T> create() {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheLoaderFactory.java:
```diff
@@ -0,0 +1,56 @@
+    return (JCacheCacheLoader<K, T>) this.instance;
+  }
+
+  public boolean equals(Object other) {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheLoaderFactory.java:
```diff
@@ -0,0 +1,56 @@
+    }
+  }
+
+  public int hashCode() {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheWriter.java:
```diff
@@ -0,0 +1,72 @@
+import java.util.Collection;
+import java.util.Iterator;
+
+public class JCacheCacheWriter<K, T extends PersistentBase> implements CacheWriter<K, T> {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheWriter.java:
```diff
@@ -0,0 +1,72 @@
+  private static final Logger LOG = LoggerFactory.getLogger(JCacheCacheWriter.class);
+  private DataStore<K, T> dataStore;
+
+  public JCacheCacheWriter(DataStore<K, T> dataStore) {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheWriter.java:
```diff
@@ -0,0 +1,72 @@
+  }
+
+  @Override
+  public void write(Cache.Entry<? extends K,
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheWriter.java:
```diff
@@ -0,0 +1,72 @@
+  }
+
+  @Override
+  public void writeAll(Collection<Cache.Entry<? extends K,
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheWriter.java:
```diff
@@ -0,0 +1,72 @@
+  }
+
+  @Override
+  public void delete(Object key) throws CacheWriterException {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheWriter.java:
```diff
@@ -0,0 +1,72 @@
+  }
+
+  @Override
+  public void deleteAll(Collection<?> keys) throws CacheWriterException {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheWriterFactory.java:
```diff
@@ -0,0 +1,55 @@
+
+import javax.cache.configuration.Factory;
+
+public class JCacheCacheWriterFactory<K, T extends PersistentBase> implements Factory<JCacheCacheWriter<K,T>> {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheWriterFactory.java:
```diff
@@ -0,0 +1,55 @@
+  public static final long serialVersionUID = 201205101621L;
+  private transient JCacheCacheWriter<K,T> instance;
+
+  public JCacheCacheWriterFactory(JCacheCacheWriter<K,T> instance) {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheWriterFactory.java:
```diff
@@ -0,0 +1,55 @@
+    this.instance = instance;
+  }
+
+  public JCacheCacheWriter<K,T> create() {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheWriterFactory.java:
```diff
@@ -0,0 +1,55 @@
+    return (JCacheCacheWriter<K,T>)this.instance;
+  }
+
+  public boolean equals(Object other) {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheCacheWriterFactory.java:
```diff
@@ -0,0 +1,55 @@
+    }
+  }
+
+  public int hashCode() {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheStore.java:
```diff
@@ -0,0 +1,412 @@
+  }
+
+  @Override
+  public void initialize(Class<K> keyClass, Class<T> persistentClass, Properties properties) {
```
Add documentation

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheStore.java:
```diff
@@ -0,0 +1,412 @@
+      this.persistentDataStore = DataStoreFactory.getDataStore(keyClass, persistentClass,
+              new Configuration());
+    } catch (GoraException ex) {
+      LOG.error("Couldn't initialize persistent DataStore");
```
Also log the Exception itself
```
 +      LOG.error("Couldn't initialize persistent DataStore", ex); 
```

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheStore.java:
```diff
@@ -0,0 +1,412 @@
+          }
+        }
+      }
+      LOG.info("JCache Gora datastore deleled " + deletedRows + " rows from Persistent datastore");
```
Use parameterized logging
http://www.slf4j.org/faq.html#logging_performance
```
 +      LOG.info("JCache Gora datastore deleled {} rows from Persistent datastore", deletedRows); 
```

lewismc commented on gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheStore.java:
```diff
@@ -0,0 +1,412 @@
+                memberOwnedCacheEntries.last(), member.getSocketAddress().getHostString());
+        partitions.add(partition);
+      }
+    } catch (java.lang.Exception ex) {
```
Is there no better exception to catch here?

lewismc commented on gora-jcache/src/test/java/org/apache/gora/jcache/store/JCacheGoraDataStoreTest.java:
```diff
@@ -0,0 +1,194 @@
+  @After
+  public void tearDown() throws Exception {
+    super.tearDown();
+    //mandatory to clean up hazelcast instances
```
Thank you for putting comments in here.

lewismc commented:
@djkevincr NICE WORK :)
Some comments
 1. Please include Javadoc for all of your publicly available methods, Classes and also add package-info.java to every package you create.
 2. Please ALWAYS catch and log Exception's within your LOG message e.g. ```LOG.error("Log message here", e) 
 3. Please ALWAYS use parameterized logging e.g. ```LOG.info("Log {} and {} values", value1, value2)```
 4. My main question is, what is going on with the testing? Are you setting up the Hazelcast test infrastructure or how is the Mock testing working? This needs to be added to the test class and needs to be explained directly there. Also why are you not extending DataStoreTestBase?
 5. As far as I can see, you've not added this to the gora-tutorial meaning that no-one can run this entire module easily. Please make sure to add it to the gora-tutorial.
 6. Please add this to gora-goraci, we will wish to test it in the Cloud shortly and may even write Docker containers for all of this.

If you can lease address the documentation and logging issues first then notify me that would be great.

Again, great work. Nice to see this coming along.

lewismc commented:
OK

lewismc commented:
The entire pull request does not compile under Java 8
When I change to Java7 it gives me the following
```
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 07:41 min
[INFO] Finished at: 2016-08-05T02:48:40-07:00
[INFO] Final Memory: 54M/323M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile (default-compile) on project gora-jcache: Compilation failure: Compilation failure:
[ERROR] /usr/local/gora-dev/gora-jcache/src/main/java/org/apache/gora/jcache/query/JCacheQuery.java:[24,7] error: cannot access Writable
[ERROR] class file for org.apache.hadoop.io.Writable not found
[ERROR] /usr/local/gora-dev/gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheStore.java:[53,29] error: package org.apache.hadoop.conf does not exist
[ERROR] /usr/local/gora-dev/gora-jcache/src/main/java/org/apache/gora/jcache/store/JCacheStore.java:[68,7] error: cannot access Configurable
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :gora-jcache
```

lewismc commented:
Thats a ```mvn clean install```

djkevincr commented:
Hi Lewis,

I haven't used any hadoop dependency over jcache module. Anyway I ll have a look into this. Can you please give some time I am still working on the adding java docs to module.

Regards
Kevin

lewismc commented:
OK, no problems. I ask you to push code as often as possible in a pull request because there are some very competent coders who can peer review your code. This is a positive thing in your GSoC endeavor.
Please push more frequently.
Please let me know how you are doing and how your code is passing tests.
We can easily verify this @ The Apache Software Foundation, we have builds. I can easily set up a branch just for your code and build your code via a branch every time you push updates.
Lets begin to be serious about integration into the master branch of Gora. This is the main goal OK.
Your work so far is excellent. We have a hurdle now, which is integration into a working, functional open source project.
You are now part of the Gora community.

lewismc commented:
BTW, in order to resolve the above conflict just look into POM

djkevincr commented:
Can you please try now?

lewismc commented:
Yes, trying right now.

lewismc commented:
@djkevincr why did you close this PR?

djkevincr commented:
@lewismc  I didn't close this pr. Did I do something mistakenly? It says this pr successfully merged and closed I do have unit tests for map reduce and several improvements to previous ones. Shall I send another pr?

lewismc commented:
@djkevincr thats fine. The build is not stable. I merged this based upon your updates. Thanks for them.

djkevincr commented:
I think the build passed without any test failures see [1] Or Am I missing anything here?
[1] https://builds.apache.org/job/gora-trunk/1674/

djkevincr commented:
@lewismc Please find PR on [1] which addresses some of your review comments as well as improvements to unit tests. 
[1] https://github.com/apache/gora/pull/73

lewismc commented:
Brilliant thanks. This has also been merged.
Please submit a PR for adding the JCache implementation to gora-tutorial. Thanks

djkevincr commented:
I will work on the gora-tutorial as well as some documentation next. Thank you.
@lewismc Seems like recent PR is not merged yet. Please find it under [1]. 
[1] https://github.com/apache/gora/pull/73
